Add Ogg Theora support to MovieWriter

berarma opened the pull request:
Supersedes #106700.

Movie Maker mode can now record files in .ogv format, which can be directly viewed in Godot's VideoStreamPlayer node along with most video players. This is a lossy format with inter-frame compression, unlike AVI + MJPEG which only performs intra-frame compression.

Thanks to @penninghlhd for the initial implementation from #98416 and @Calinou for the cleanup in #106700. 🙂

What I've done:

- Fix issues in the generated video related to the multiplexing of video and audio in an Ogg container. Reported by `oggz-validate` and debugged with `oggz-dump`.
- Fix compilation when using builtin_libtheora=no builtin_libvorbis=no builtin_libogg=no (undefined references).
- Move MovieWriterOGV to the modules/theora module it depends on, like https://github.com/godotengine/godot/pull/106013.
- Add more project settings to configure the video encoding process.
- Fix video output by converting the image to RGB8 when it's in a different format. The issue can be reproduced with [this demo](https://godotengine.org/asset-library/asset/2762).

I've added many settings, but it can help testing them and decide which ones we want.  In the case of `video_bitrate`, it enables CBR and it might provide better encoding speed in some cases. Would it be really bad to provide all these options?

## TODO

- [ ] Test on Windows + MSVC and make sure the Theora files are correct once they're fixed on other platforms. MSVC uses a different set of files for x86 intrinsics in Theora. (Help needed on this one.)

The message `[ffmpeg/demuxer] ogg: Broken file, keyframe not correctly marked.` in FFmpeg is still showing, but it's not really an error and it shouldn't affect operations. I think it's just FFmpeg being finicky.

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1105,21 +1105,34 @@
+			Forces keyframes at the specified interval. Higher values can improve compression up to a certain level at the expense of some seek latency.
+		</member>
+		<member name="editor/movie_writer/ogv/video_bitrate" type="int" setter="" getter="" default="0">
+			When greater than [code]0[/code] CBR is used instead of VBR and sets the target bitrate for the video encoder in KB. The higher the value the better video quality and bigger file size. VBR should give better results and it's thus recommended to leave this value at [code]0[/code].
```
```suggestion
			When greater than [code]0[/code], a constant bitrate (CBR) is used instead of variable bitrate (VBR). The value sets the target bitrate for the video encoder in kilobytes per second. The higher the value, the better video quality and the larger the file size. In most cases, VBR will give better results. Therefore, it's recommended to leave this value at [code]0[/code].
```

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1105,21 +1105,34 @@
+			[b]Note:[/b] This does not affect video quality.
+		</member>
+		<member name="editor/movie_writer/ogv/encoding_speed" type="int" setter="" getter="" default="3">
+			Trades speed for compression. Speed [code]1[/code] is the slowest but provides the best compression. Speed [code]4[/code] is the fastest but provides the worst compression.
```
```suggestion
			The tradeoff between speed and compression efficiency. Speed [code]1[/code] is the slowest but provides the best compression. Speed [code]4[/code] is the fastest but provides the worst compression. Video quality is generally not affected significantly by this setting.
```

Calinou commented on servers/movie_writer/movie_writer.cpp:
```diff
@@ -211,11 +217,18 @@ void MovieWriter::add_frame() {
 }
 
 void MovieWriter::end() {
+	int encoding_start = Time::get_singleton()->get_ticks_usec();
 	write_end();
+	int encoding_end = Time::get_singleton()->get_ticks_usec();
+	encoding_time += encoding_end - encoding_start;
```
```suggestion
	uint64_t encoding_start_usec = Time::get_singleton()->get_ticks_usec();
	write_end();
	uint64_t encoding_end_usec = Time::get_singleton()->get_ticks_usec();
	encoding_time_usec += encoding_end_usec - encoding_start_usec;
```

Calinou reviewed (commented):
Tested locally, it works as expected. Code and documentation looks good to me. I like the addition of the encoding time metric :slightly_smiling_face: 

**Testing project:** https://github.com/Calinou/godot-movie-maker-demo

Example output file with the default settings:

```
Input #0, ogg, from 'out.ogv':
  Duration: 00:00:30.00, start: 0.000000, bitrate: 27760 kb/s
  Stream #0:0: Video: theora, yuv420p, 1280x720 [SAR 1:1 DAR 16:9], 60 fps, 60 tbr, 60 tbn
  Stream #0:1: Audio: vorbis, 48000 Hz, stereo, fltp, 96 kb/s
```

The file plays back correctly in mpv.

Encoding speeds (measured on a debug build):

### OGV

```
Done recording movie at path: /tmp/out.ogv
1800 frames at 60 FPS (movie length: 00:00:30:00), recorded in 00:01:41 (29% of real-time speed).
CPU time: 1.40 seconds (average: 0.78 ms/frame)
GPU time: 10.53 seconds (average: 5.85 ms/frame)
Encoding time: 30.79 seconds (average: 17.10 ms/frame)
```

### AVI

```
Done recording movie at path: /tmp/out.avi
1800 frames at 60 FPS (movie length: 00:00:30:00), recorded in 00:01:35 (31% of real-time speed).
CPU time: 1.39 seconds (average: 0.77 ms/frame)
GPU time: 9.57 seconds (average: 5.32 ms/frame)
Encoding time: 28.88 seconds (average: 16.04 ms/frame)
```

### PNG/WAV

```
Done recording movie at path: /tmp/out.png
1800 frames at 60 FPS (movie length: 00:00:30:00), recorded in 00:04:01 (12% of real-time speed).
CPU time: 1.56 seconds (average: 0.87 ms/frame)
GPU time: 33.50 seconds (average: 18.61 ms/frame)
Encoding time: 148.37 seconds (average: 82.43 ms/frame)
```

You can add an `OGV` capitalization here so the Project Settings category uses correct casing: https://github.com/godotengine/godot/blob/26df04377e897595495980309535064e1bacdf53/editor/editor_property_name_processor.cpp#L247

Calinou commented:
> Using CBR with a target bitrate equal to the mean bitrate of the VBR file for equivalent compression takes almost the same time as VBR. It's even slower at speed 1. This setting might be unnecessary.

I agree, I suggest removing the CBR setting if we can't find a good use case for it.

Calinou commented on servers/movie_writer/movie_writer.cpp:
```diff
@@ -155,7 +156,12 @@ void MovieWriter::_bind_methods() {
-	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/mjpeg_quality", PROPERTY_HINT_RANGE, "0.01,1.0,0.01"), 0.75);
+	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/video_quality", PROPERTY_HINT_RANGE, "0.0,1.0,0.01"), 0.75);
+	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/ogv/audio_quality", PROPERTY_HINT_RANGE, "-0.1,1.0,0.01"), 0.2);
+	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/encoding_speed", PROPERTY_HINT_ENUM, "1,2,3,4"), 3);
```
```suggestion
	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/encoding_speed", PROPERTY_HINT_ENUM, "Fastest (Lowest Efficiency):3,Fast (Low Efficiency):2,Slow (High Efficiency):1,Slowest (Highest Efficiency):0"), 3);
```

This allows the project setting to show up like this, keeping compatibility with the existing enum:

![image](https://github.com/user-attachments/assets/de6991e3-8992-4ef5-887b-694ae34687a4)

berarma replied:
Yes, I had thought about this, but didn't get to put it into words. Thanks!

berarma commented on servers/movie_writer/movie_writer.cpp:
```diff
@@ -244,5 +257,6 @@ void MovieWriter::end() {
 	print_line(vformat("%d frames at %d FPS (movie length: %s), recorded in %s (%d%% of real-time speed).", Engine::get_singleton()->get_frames_drawn(), fps, movie_time, real_time, (float(MAX(1, movie_time_seconds)) / MAX(1, real_time_seconds)) * 100));
 	print_line(vformat("CPU time: %.2f seconds (average: %.2f ms/frame)", cpu_time / 1000, cpu_time / Engine::get_singleton()->get_frames_drawn()));
 	print_line(vformat("GPU time: %.2f seconds (average: %.2f ms/frame)", gpu_time / 1000, gpu_time / Engine::get_singleton()->get_frames_drawn()));
```
Perfect.

berarma commented:
Thanks for all the suggestions, and for saving me from my struggles when writing documentation.

I've removed the `video_bitrate` setting, renumbered the speed labels to use the real value instead of being off by one, and capitalized the OGV label (I was hoping for something like this to exist :) ).

And rebased it on top of the current master, now with Theora 1.2.0.

I forgot to mention it before but I'm using `oggz-validate` to check the generated OGV files. It reported errors for Ogg pages with decreasing timestamps and the lack of end of stream markers. Now the generated files are clean.

berarma committed:
Add Ogg Theora support to MovieWriter

Movie Maker mode can now record files in `.ogv` format, which can be
directly viewed in Godot's VideoStreamPlayer node along with most
video players. This is a lossy format with inter-frame compression,
unlike AVI + MJPEG which only performs intra-frame compression.

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>
Co-authored-by: Leo de Penning <leo.depenning@illuminoo.com>

berarma commented:
I've put everything in one commit.

akien-mga commented on modules/theora/movie_writer_ogv.cpp:
```diff
@@ -0,0 +1,428 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "movie_writer_ogv.h"
+#include "core/config/project_settings.h"
+#include "thirdparty/misc/rgb2yuv.h"
```
```suggestion
#include "movie_writer_ogv.h"

#include "core/config/project_settings.h"

#include "thirdparty/misc/rgb2yuv.h"
```

Following https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#header-includes

akien-mga commented on servers/movie_writer/SCsub:
```diff
@@ -4,3 +4,11 @@ from misc.utility.scons_hints import *
 Import("env")
 
 env.add_source_files(env.servers_sources, "*.cpp")
+
+# Also requires libogg, libtheora and libvorbis.
+if env["builtin_libogg"]:
+    env.Prepend(CPPPATH=["#thirdparty/libogg"])
+if env["builtin_libtheora"]:
+    env.Prepend(CPPPATH=["#thirdparty/libtheora", "#thirdparty/misc"])
+if env["builtin_libvorbis"]:
+    env.Prepend(CPPPATH=["#thirdparty/libvorbis"])
```
This seems wrong, those should only be needed for `movie_writer_ogv.h` which is in the `theora` module.

akien-mga commented on thirdparty/misc/rgb2yuv.h:
```diff
@@ -0,0 +1,41 @@
+#ifndef RGB2YUV_H
+#define RGB2YUV_H
+
+#include "core/typedefs.h"
```
The source of this file needs to be documented in `thirdparty/README.md` and its copyright in `COPYRIGHT.txt`.

If you wrote it yourself, it shouldn't be in `thirdparty`.

berarma replied:
I inherited it from the original PR. It looks like it comes [stackoverflow](https://stackoverflow.com/a/9467305). I'll document it.

berarma replied:
Please, see comment in the PR.

akien-mga commented on modules/theora/SCsub:
```diff
@@ -22,49 +22,49 @@ if env["builtin_libtheora"]:
-        # "encfrag.c",
-        # "encinfo.c",
-        # "encode.c",
+        "encfrag.c",
+        "encinfo.c",
+        "encode.c",
```
Is there any use case for having the theora encoder at runtime in export templates?

If we need it only for the editor, we should probably gate compiling these files (and adding the `theoraenc` and `vorbisenc` dependencies in `linuxbsd/detect.py`) on `env.editor_build`.

This helps keep export templates compat.

berarma commented:
I've tried to address the issues, although I have doubts about the inclusion of CC-BY-SA 3.0 code.

I've made a few minor changes to the `rgb2yuv` that could improve performance slightly and make the code simpler. Should I add a comment to the file about this?

I've added a check for the video dimensions because the encoding won't work if they aren't even.

berarma commented:
@akien-mga, it looks like it's trying to build the templates with the MovieWriter enabled, and it can't find the functions in the files that I've just moved to editor builds only.

AThousandShips commented on modules/theora/movie_writer_ogv.cpp:
```diff
@@ -0,0 +1,431 @@
+	vorbis_analysis_wrote(&vd, audio_frames);
+}
+
+void MovieWriterOGV::pull_audio(const bool last) {
```
```suggestion
void MovieWriterOGV::pull_audio(bool p_last) {
```

AThousandShips commented on modules/theora/movie_writer_ogv.cpp:
```diff
@@ -0,0 +1,431 @@
+	th_encode_ycbcr_in(td, ycbcr);
+}
+
+void MovieWriterOGV::pull_video(const bool last) {
```
```suggestion
void MovieWriterOGV::pull_video(bool p_last) {
```

AThousandShips commented on modules/theora/movie_writer_ogv.cpp:
```diff
@@ -0,0 +1,431 @@
+// The added complexity here is because we have to ensure pages are written in ascending timestamp order.
+// libOgg doesn't allow checking the next page granulepos without requesting the page, and once requested it can't be
+// returned, thus, we need to save it so that it doesn't get erased by the next `ogg_stream_packetin` call.
+void MovieWriterOGV::write_to_file(const bool finish) {
```
```suggestion
void MovieWriterOGV::write_to_file(bool p_finish) {
```

akien-mga reviewed (commented):
Looks pretty good to me!

I notice that, unlike what I might have suggested before, `--movie-writer` seems to be exposed in non-editor builds (export templates). I don't know if it actually works that way, but that means that with the changes done to only include the theora encoder in editor builds, OGV won't be available as a format for runtime movie encoding.

I think it's an appropriate tradeoff for now, and we can reconsider it if there's significant demand and the binary size impact is not too big.

berarma commented on thirdparty/README.md:
```diff
@@ -828,6 +828,10 @@ Collection of single-file libraries used in Godot components.
   * Upstream: http://wss.co.uk/pinknoise/yuv2rgb/ (to check)
   * Version: ?
   * License: BSD
+- `rgb2yuv.h`
+  * Upstream: https://stackoverflow.com/a/9467305
+  * Version: 2012-02-27
+  * License: CC-BY-SA 3.0
```
I agree.

berarma commented on servers/movie_writer/movie_writer.cpp:
```diff
@@ -155,7 +156,11 @@ void MovieWriter::_bind_methods() {
 	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/speaker_mode", PROPERTY_HINT_ENUM, "Stereo,3.1,5.1,7.1"), 0);
-	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/mjpeg_quality", PROPERTY_HINT_RANGE, "0.01,1.0,0.01"), 0.75);
+	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/video_quality", PROPERTY_HINT_RANGE, "0.0,1.0,0.01"), 0.75);
+	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/ogv/audio_quality", PROPERTY_HINT_RANGE, "-0.1,1.0,0.01"), 0.5);
+	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/encoding_speed", PROPERTY_HINT_ENUM, "Fastest (Lowest Efficiency):4,Fast (Low Efficiency):3,Slow (High Efficiency):2,Slowest (Highest Efficiency):1"), 4);
+	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/keyframe_interval", PROPERTY_HINT_RANGE, "1,1024,1"), 64);
```
The other movie formats don't do this, so I couldn't use them as example.

I've tried to put them in a `_bind_methods` function in MovieWriterOGV and it removes the settings when the module is disabled, but I get several virtual method errors at runtime on methods that I haven't touched:

```
WARNING: Property not found: 'editor/movie_writer/ogv/audio_quality'.
     at: get_setting_with_override (core/config/project_settings.cpp:404)
WARNING: Property not found: 'editor/movie_writer/ogv/encoding_speed'.
     at: get_setting_with_override (core/config/project_settings.cpp:404)
WARNING: Property not found: 'editor/movie_writer/ogv/keyframe_interval'.
     at: get_setting_with_override (core/config/project_settings.cpp:404)
ERROR: Class 'MovieWriter' already has a method 'add_writer'.
   at: bind_methodfi (core/object/class_db.cpp:1946)
ERROR: Virtual method already bound 'MovieWriter::_get_audio_mix_rate'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
ERROR: Virtual method already bound 'MovieWriter::_get_audio_speaker_mode'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
ERROR: Virtual method already bound 'MovieWriter::_handles_file'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
ERROR: Virtual method already bound 'MovieWriter::_write_begin'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
ERROR: Virtual method already bound 'MovieWriter::_write_frame'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
ERROR: Virtual method already bound 'MovieWriter::_write_end'.
   at: add_virtual_method (core/object/class_db.cpp:2024)
```

akien-mga replied:
My suggestion was just to do something like this:
```diff
diff --git a/modules/theora/register_types.cpp b/modules/theora/register_types.cpp
index ee2ac93593..723221b2a7 100644
--- a/modules/theora/register_types.cpp
+++ b/modules/theora/register_types.cpp
@@ -36,6 +36,8 @@
 #include "editor/movie_writer_ogv.h"
 #endif
 
+#include "core/config/project_settings.h"
+
 static Ref<ResourceFormatLoaderTheora> resource_loader_theora;
 #ifdef TOOLS_ENABLED
 static MovieWriterOGV *writer_ogv = nullptr;
@@ -46,6 +48,10 @@ void initialize_theora_module(ModuleInitializationLevel p_level) {
 		case MODULE_INITIALIZATION_LEVEL_SERVERS: {
 #ifdef TOOLS_ENABLED
 			if (GD_IS_CLASS_ENABLED(MovieWriterOGV)) {
+				GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/ogv/audio_quality", PROPERTY_HINT_RANGE, "-0.1,1.0,0.01"), 0.5);
+				GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/encoding_speed", PROPERTY_HINT_ENUM, "Fastest (Lowest Efficiency):4,Fast (Low Efficiency):3,Slow (High Efficiency):2,Slowest (Highest Efficiency):1"), 4);
+				GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/keyframe_interval", PROPERTY_HINT_RANGE, "1,1024,1"), 64);
+
 				writer_ogv = memnew(MovieWriterOGV);
 				MovieWriter::add_writer(writer_ogv);
 			}
@@ -57,6 +63,7 @@ void initialize_theora_module(ModuleInitializationLevel p_level) {
 			ResourceLoader::add_resource_format_loader(resource_loader_theora, true);
 			GDREGISTER_CLASS(VideoStreamTheora);
 		} break;
+
 		default:
 			break;
 	}
diff --git a/servers/movie_writer/movie_writer.cpp b/servers/movie_writer/movie_writer.cpp
index bba7c56c9f..f22253bd44 100644
--- a/servers/movie_writer/movie_writer.cpp
+++ b/servers/movie_writer/movie_writer.cpp
@@ -157,9 +157,6 @@ void MovieWriter::_bind_methods() {
 	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/mix_rate", PROPERTY_HINT_RANGE, "8000,192000,1,suffix:Hz"), 48000);
 	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/speaker_mode", PROPERTY_HINT_ENUM, "Stereo,3.1,5.1,7.1"), 0);
 	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/video_quality", PROPERTY_HINT_RANGE, "0.0,1.0,0.01"), 0.75);
-	GLOBAL_DEF(PropertyInfo(Variant::FLOAT, "editor/movie_writer/ogv/audio_quality", PROPERTY_HINT_RANGE, "-0.1,1.0,0.01"), 0.5);
-	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/encoding_speed", PROPERTY_HINT_ENUM, "Fastest (Lowest Efficiency):4,Fast (Low Efficiency):3,Slow (High Efficiency):2,Slowest (Highest Efficiency):1"), 4);
-	GLOBAL_DEF(PropertyInfo(Variant::INT, "editor/movie_writer/ogv/keyframe_interval", PROPERTY_HINT_RANGE, "1,1024,1"), 64);
 
 	// Used by the editor.
 	GLOBAL_DEF_BASIC("editor/movie_writer/movie_file", "");
```

I don't think this impacts users, the settings are still registered properly and available in the editor. The only difference is that if the theora module is disabled at compile, then those settings aren't exposed anymore.

Which IMO makes sense because they're a no-op in this case. One potential drawback if people go back and forth between theora-enabled and disabled build is that they could lose some config, but that doesn't seem like a likely scenario.

berarma replied:
Sorry, I didn't get to see your messages until now. They probably arrived while I was writing mine and didn't see any notifications. Marking as solved since the PR is already merged.

I didn't know it could be done like that. I was overriding the `_bind_methods` method and that got me the virtual method errors.  Well, maybe for another PR.

Thanks for the help.

berarma commented on modules/theora/editor/movie_writer_ogv.cpp:
```diff
@@ -0,0 +1,431 @@
+/**************************************************************************/
+
+#include "movie_writer_ogv.h"
+#include "rgb2yuv.h"
```
Ah, sorry, I'm still getting used to that rule. I've also removed the line above to join it to the following include and the formatter has sorted them.

akien-mga commented on modules/theora/editor/rgb2yuv.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef RGB2YUV_H
+#define RGB2YUV_H
```
This needs the Godot copyright header and switching to `#pragma once`:
https://github.com/godotengine/godot/actions/runs/15560984396/job/43812941752?pr=107188

You can do it easily with `pre-commit`:

```
pip install pre-commit
pre-commit install
pre-commit run --all
```

berarma commented:
> I notice that, unlike what I might have suggested before, `--movie-writer` seems to be exposed in non-editor builds (export templates). I don't know if it actually works that way, but that means that with the changes done to only include the theora encoder in editor builds, OGV won't be available as a format for runtime movie encoding.

I didn't think to check it out. It made sense that it was editor-only.

> I think it's an appropriate tradeoff for now, and we can reconsider it if there's significant demand and the binary size impact is not too big.

Don't hesitate to ask me if you change your mind now or later, it's easy to revert.

akien-mga reviewed (approved):
Looks great to me!
